docs/csharp/advanced-topics/performance/index.md

```diff
@@ -0,0 +1,94 @@
+
+In C#, parameters to methods are *passed by value*, and return values are *return by value*. The *value* of the argument is passed to the method. The *value* of the return argument is the return value.
+
+The `ref`, `in`, or `out` modifier indicates that parameter is *passed by reference*. The *reference* to the storage location is passed to the method. Adding `ref` to the method signature means the return value is *returned by reference*. The *reference* to the storage location is the return value.
+
+You can also use *ref assignment* to have a variable refer to another variable. A typical assignment copies the *value* of the right hand side to the variable on the left hand side of the assignment. A *ref assignment* copies the memory location of the variable on the right hand side to the variable on the left hand side. The `ref` now refers to the original variable:
+
+:::code language="csharp" source="./snippets/ref-safety/Program.cs" id="RefAssignment":::
+
+When you *assign* a variable, you change its *value*. When you *ref assign* a variable, you change what it refers to.
+
+You can work directly with the storage for values using `ref` variables, pass by reference, and ref assignment. Scope rules enforced by the compiler ensure safety when working directly with storage.
+
+## Ref safe to escape scope
+
+C# includes rules for `ref` expressions to ensure that a `ref` expression can't be accessed where the storage it refers to is no longer valid. Consider the following example:
+
+```csharp
+public ref int CantEscape()
+{
+    int index = 42;
+    return ref index; // Error: error's ref safe to escape scope is the body of CantEscape
+}
+```
+
+The compiler reports an error because you can't return a reference to a local variable from a method. The caller can't access the storage being referred to. The *ref safe to escape scope* defines the scope in which a `ref` expression is safe to access or modify. For non-ref locals, the *ref safe to escape* scope is the body of the method. In other words, a non-ref local can't be returned as a `ref` expression. For a variable to have a *ref safe to escape scope* of the entire method, it must be accessible and assignable outside the method. The following snippet shows two examples. A member field can be accessed from the scope calling a method, so a class or struct field's *ref safe to escape scope* is the entire method. The *ref safe to escape scope* for a parameter with the `ref`, `in`, or `out` modifiers is also the entire method. Both can be `ref` returned from a member method:
```

Comment:
>Both can be `ref` returned from a member method: [](http://example.com/codeflow?start=844&length=48)

The _ref safe to escape scope_ of a class field is the calling method, so a ref to a class field can be returned from a member method. But the _ref safe to escape scope_ of a non-ref struct field is the current method, so a ref to a struct field cannot be returned.

Reply:
For parameters, in C#11 the _ref safe to escape scope_ of an `out` parameter is the current method, so an `out` parameter cannot be returned by reference.

Reply:
I restructured this paragraph as a table that states the *ref safe to escape* scopes for all different declaration types, based on these notes. I think I got it correct, but it's worth another look.